Issue 571: Require completion of a survey before yearly allowance may be renewed

haudquaquam opened the pull request:
Fixes #571 

**Changes**

- Added a JSON field, `renewal_survey_answers` to `AllocationRenewalRequest` model to store survey responses. Default value is an empty dict.
- Introduced a new renewal form, `ProjectRenewalSurveyForm` (modeled after `SavioProjectSurveyForm`): 
  - Checks for `BRC_ONLY` or `LRC_ONLY` flag at runtime to determine which questions to display. BRC and LRC will use different survey questions.
  - Current questions are placeholders and should be updated accordingly.
  - Added template `project_renewal_survey.html`, modeled after `new_project_survey.html`.
  - Imported the form and template into the two views through which a user can request an allocation renewal: `AllocationRenewalRequestUnderProjectView` and `AllocationRenewalRequestView` in `renewal_views/request_views.py`.

**How to Test**

- Manual Testing
  - As a user who is on a project, go to the project's detail page -> _Renew Allowance_, or go to _Project -> Projects -> Renew a PI's Allowance_.
  - Complete the renewal request flow, and ensure the survey is there and has the desired questions depending on if the deployment is LRC or BRC.
  - After submitting the renewal request, go to the Coldfront Administration dashboard and view the newly created Allocation Renewal Request object. Survey answers should appear in JSON format under **Renewal survey answers**.

matthew-li commented on coldfront/core/project/views_/renewal_views/request_views.py:
```diff
@@ -128,7 +129,7 @@ def __init__(self, *args, **kwargs):
     @staticmethod
     def create_allocation_renewal_request(requester, pi, computing_allowance,
-                                          allocation_period, pre_project,
+                                          allocation_period, renewal_survey_answers, pre_project,
```
Minor nitpick: I'd put this new parameter after `pre_project` and `post_project` to match the ordering of the form.

haudquaquam replied:
Done

matthew-li commented:
Thanks, code changes so far LGTM. I'll take a closer look when the actual survey has been integrated.

(Reminder to add a popup to the renewal request detail view, showing the answers.)

haudquaquam commented:
**Changes**

- Added BRC survey questions that will be used for the upcoming allocation renewal request period.
- Added subcommand `renewal_survey_responses` to `export_data` command:
  - Changed original `survey_responses` to `new_project_responses`
  - `renewal_survey_responses` is a subcommand that will print the responses of `AllocationRenewalRequest` renewal survey responses given a format and allocation period.
    - Raises an exception if the given allocation period does not have a required survey (this is done by checking if any of the surveys of the `AllocationRenewalRequest` objects within a certain `allocation_period` are empty.)
    - Requires `allocation_period` flag where user specifies the period (i.e. 'Allowance Year 2024 - 2025'.)
    - Prints out human-readable question and answer names instead of default form IDs.

**How to Test**

- Manual Testing
  - Create an `AllocationRenewalRequest` with the corresponding survey answers under the 2024-2025 allowance year (since the previous years did not have required renewal forms.)
  - Run `python3 manage.py export_data renewal_survey_responses --allocation_period 'Allowance Year 2024 - 2025' --format json`
  - Change format to `csv` and ensure that it displays properly as well.
  - Try earlier allocation periods (i.e. `'Allowance Year 2023 - 2024'`) to make sure that the proper exception is thrown, indicating that allocation periods earlier than 2024 - 2025 did not have required renewal surveys.

matthew-li commented on coldfront/core/utils/management/commands/export_data.py:
```diff
@@ -131,14 +135,29 @@ def add_subparsers(self, subparsers):
+                                                           help='Export survey responses for project renewals')
+        renewal_survey_responses_subparser.add_argument('--format', help='Format to dump renewal survey responses in',
+                                                type=str, required=True, choices=['json', 'csv'])
+        renewal_survey_responses_subparser.add_argument('--allocation_period',
```
Getting an error for some periods:

```
python3 manage.py export_data renewal_survey_responses --format csv --allocation_period "Allowance Year 2020 - 2021"
list index out of range
```

```
python3 manage.py export_data renewal_survey_responses --format csv --allocation_period "Allowance Year 2024 - 2025"
list index out of range
```

haudquaquam replied:
Thanks Matthew, I'm seeing that these allocation periods have no renewal requests with surveys, hence the `list index out of range` error. I'm adding a step that throws an exception in this case now

matthew-li commented on coldfront/core/utils/management/commands/export_data.py:
```diff
@@ -452,7 +471,95 @@ def handle_survey_responses(self, *args, **kwargs):
+                            survey[question] = [sub_map.get(i,i) for i in answer]
+                        else:
+                            # Single choice replacement
+                            survey[question] = sub_map[answer]
```
```
  File "/var/www/coldfront_app/coldfront/coldfront/core/utils/management/commands/export_data.py", line 525, in handle_renewal_survey_responses
    survey[question] = sub_map[answer]
KeyError: ''
```

Didn't investigate deeply, but I think this is happening because the multiple choice fields (mentioned below) are not required.

haudquaquam replied:
Fixed by adding a check to skip swapping for empty answers

matthew-li reviewed (changes requested):
Thanks for adding this! Overall LGTM.

I've added some comments, but nothing major. We're also separately discussing whether some of these fields should be required.

haudquaquam commented on coldfront/core/project/forms_/renewal_forms/request_forms.py:
```diff
@@ -191,6 +194,382 @@ def __init__(self, *args, **kwargs):
+                required=False,
+                widget=forms.Textarea(attrs={'rows': 3}))
+            
+            self.fields['indicate_topic_interests'] = forms.MultipleChoiceField(
```
Added a 'none of the above' option

haudquaquam commented:
**Changes**
- Made several small changes based on @matthew-li's comments (i.e. grouping imports, removing leftover debugging code, renaming variables, tweaking BRC renewal survey questions).
- `export_data` renewal survey responses function:
  - Fixed bugs covering empty answers and `list index out of bounds` for invalid allocation periods.
  - Extracted logic for swapping answer IDs for plain-text into `_swap_form_answer_id_for_text`.
- Added steps in allocation renewal request tests to fill in the renewal survey.

haudquaquam commented on coldfront/core/project/forms_/renewal_forms/request_forms.py:
```diff
@@ -191,6 +194,382 @@ def __init__(self, *args, **kwargs):
+                required=True,
+                widget=forms.CheckboxSelectMultiple())
+
+            self.fields['training_session_usefulness_of_computational_platforms_training'] = forms.ChoiceField(
```
Leaving as is for now, don't think an N/A option would be necessary as that seems equivalent to choosing '1'

matthew-li commented:
Changes:
- Got rid of the merge migration.
- Added temporary logic to ensure that the survey form only appears for "Allowance Year 2024 - 2025". This will need to be updated before the subsequent period.
- Added tests for the above logic.
- Updated the `app-base` Docker image to install `libfaketime`, which can be used to mock a date/time by setting some environment variables.

matthew-li reviewed (approved):
Thanks for putting this together.

- I've added the aforementioned logic to ensure the survey only appears for the upcoming period and not the current one.
- I didn't run into any issues when testing on BRC or LRC instances.
- I'll review the `export_data` command later, when it's relevant.
